Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,860 @@
+   *
+   * \param[in] mpi_env - MPI environment for communication
+   * \param[out] Nodes_all - [k][r] is set to number of nodes in layer (k+1), rank r.
+   * \param[in] local_version - local instance to be send to the other processes
+   */
+  void count_g(Nd_MPI_Environment const& mpi_env, su2matrix<Index_t>& Nodes_all, CurrentLayer const& local_version) {
+    nNodes = 0;
+    // gather numbers of nodes in the current layer from all processes
+    mpi_env.MPI_Allgather_fun(&(local_version.nNodes), 1, mpi_env.mpi_index, Nodes_all[K - 1], 1, mpi_env.mpi_index,
+                              mpi_env.comm);
+    for (int r = 0; r < mpi_env.size; r++) {
+      nNodes += Nodes_all[K - 1][r];
+    }
+    Base::count_g(mpi_env, Nodes_all, local_version);
+  }
+
+  /*! \brief Gather the distributed flatteners' data and index arrays into the allocated arrays.
+   *
+   * \param[in] mpi_env - MPI environment for communication
+   * \param[in] Nodes_all - [k][r] is the number of nodes in layer (k+1), rank r.
+   * \param[in] local_version - local instance to be sent to the other processes
+   */
+  void set_g(Nd_MPI_Environment const& mpi_env, su2matrix<Index_t> const& Nodes_all,
+             CurrentLayer const& local_version) {
+    std::vector<int> Nodes_all_K_as_int(mpi_env.size);
+    std::vector<int> Nodes_all_k_cumulated( mpi_env.size + 1);
+    //< [r] is the number of nodes in the current layer, summed over all processes with rank below r, **plus one**.
+    // Used as displacements in Allgatherv, as we do not want to transfer the initial zeros, but we want to transfer the
+    // last element of indices, which is the local nNodes of the layer below. Note that MPI needs indices of type 'int'.
+    Nodes_all_k_cumulated[0] = 1;
+    for (int r = 0; r < mpi_env.size; r++) {
+      Nodes_all_k_cumulated[r + 1] = Nodes_all_k_cumulated[r] + Nodes_all[K - 1][r];
+      Nodes_all_K_as_int[r] = Nodes_all[K - 1][r];
+    }
+    std::cout << "indices[0] before Allgatherv: " << indices[0] << "\n";
+    mpi_env.MPI_Allgatherv_fun(local_version.indices.data() + 1, Nodes_all[K - 1][mpi_env.rank], mpi_env.mpi_index,
+                               indices.data(), Nodes_all_K_as_int.data(), Nodes_all_k_cumulated.data(),
+                               mpi_env.mpi_index, mpi_env.comm);
+    std::cout << "indices[0] after Allgatherv: " << indices[0] << "\n";
+
+    std::cout << "parameter to MPI_Allgatherv set_g K="<<K<<" were: "
+             << local_version.indices.data()+1 << ";"
+             << Nodes_all[K - 1][mpi_env.rank] << ";"
+             <<mpi_env.mpi_index << ";"
+             <<indices.data()<<";"
+             <<Nodes_all_K_as_int.data()<<";"
+             <<Nodes_all_k_cumulated.data()<<";"
+             <<mpi_env.mpi_index<<";"
+             << mpi_env.comm<<"\n";
+    std::cout << "here, Nodes_all_K_as_int[0]="<<Nodes_all_K_as_int[0]<<", Nodes_all_k_cumulated[0]="<<Nodes_all_k_cumulated[0]<<" size="<<mpi_env.size<<"\n";
```

Comment:
CI output for serial AD discadj_incomp_cylinder:

    indices[0] before Allgatherv: 0
    indices[0] after Allgatherv: 2
    parameter to MPI_Allgatherv set_g K=2 were: 0x560568495898;1;1;0x560568846750;0x56056889d270;0x560568ba2210;1;0
    here, Nodes_all_K_as_int[0]=1, Nodes_all_k_cumulated[0]=1 size=1

indices[0] should remain 0...